tests/test_proxy_use.py

```diff
@@ -0,0 +1,366 @@
+    # start listening before allowing tests to begin, lest we get "Connection
+    # refused" errors. On the first test system. 0.1s was too short and 0.15s
+    # was long enough. Use 0.5s to be safe, and if issues arise, increase it.
+    time.sleep(0.5)
+
+
+
+
+
+  @classmethod
+  def tearDownClass(cls):
+    """
+    Cleanup performed after the last of the tests (TestWithProxies methods)
+    has been run.
+    Stop server process and perform clean up.
+    """
+    unittest_toolbox.Modified_TestCase.tearDownClass()
+
+    for proc in [
+        cls.http_server_proc,
+        cls.https_server_proc,
+        cls.http_proxy_proc,
+        cls.https_proxy_proc,
+      ]:
+      if proc.returncode is None:
+        logger.info('\tTerminating process ' + str(proc.pid) + ' in cleanup.')
+        proc.kill()
+
+
+
+  def setUp(self):
+    """
+    Setup performed before EACH test function (TestWithProxies class method)
+    runs.
+    """
+    unittest_toolbox.Modified_TestCase.setUp(self)
+
+    # Dictionary for saving environment values to restore.
+    self.old_env_values = {}
+
+    # Make a temporary file to serve on the server, and determine its length,
+    # and its url on the server.
+    current_dir = os.getcwd()
+    target_filepath = self.make_temp_data_file(directory=current_dir)
+    rel_target_filepath = os.path.basename(target_filepath)
+
+    with open(target_filepath, 'r') as target_file_object:
+      self.target_data_length = len(target_file_object.read())
+
+    self.url = \
+        'http://localhost:' + str(self.http_port) + '/' + rel_target_filepath
+
+    self.url_https = \
+        'https://localhost:' + str(self.https_port) + '/' + rel_target_filepath
+
+
+
+
+
+  def tearDown(self):
+    """
+    Cleanup performed after each test (each TestWithProxies method).
+    Reset environment variables (for next test, etc.).
+    """
+    unittest_toolbox.Modified_TestCase.tearDown(self)
+
+    self.restore_all_modified_env_values()
+
+
+
+
+
+  def test_baseline_no_proxy(self):
+    """
+    Test a length-validating TUF download of a file through a proxy. Use an
+    HTTP proxy, and perform an HTTP connection with the final server.
+    """
+
+    logger.info('Trying http download with no proxy: ' + self.url)
+    download.safe_download(self.url, self.target_data_length)
+    download.unsafe_download(self.url, self.target_data_length)
+
+
+
+
+
+  def test_http_dl_via_smart_http_proxy(self):
+    """
+    Test a length-validating TUF download of a file through a proxy. Use an
+    HTTP proxy normally, and make an HTTP connection with the final server.
+    """
+
+    self.set_env_value('HTTP_PROXY', self.http_proxy_addr)
+
+    logger.info('Trying http download via http proxy: ' + self.url)
+    download.safe_download(self.url, self.target_data_length)
+    download.unsafe_download(self.url, self.target_data_length)
+
+
+
+
+
+  def test_httpS_dl_via_smart_http_proxy(self):
```

Comment:
(I think I get the idea of capitalization ~ highlighting, but this looks terrible to me.)